vsintegration/src/FSharp.Editor/CodeLens/LineLens.fs

```diff
@@ -0,0 +1,697 @@
+            if self.RelayoutRequested.Count > 0 then
+                self.RelayoutRequested.Dequeue() |> ignore
+                for lineNumber in visibleLineNumbers do
+                    let line = 
+                        (buffer.GetLineFromLineNumber lineNumber).Start
+                        |> view.GetTextViewLineContainingBufferPosition
+                    applyFuncOnLineStackPanels line (fun ui ->
+                        ui.Visibility <- Visibility.Visible
+                        layoutUIElementOnLine line ui
+                    )
+            // Save the new first and last visible lines for tracking
+            recentFirstVsblLineNmbr <- firstVisibleLineNumber
+            recentLastVsblLineNmbr <- lastVisibleLineNumber
+            // We can cancel existing stuff because the algorithm supports abortion without any data loss
+            layoutChangedCts.Cancel()
+            layoutChangedCts.Dispose()
+            layoutChangedCts <- new CancellationTokenSource()
+            asyncMaybe {
+                // Suspend 16 ms, instantly applying the layout to the adornment elements isn't needed 
+                // and would consume too much performance
+                do! Async.Sleep(16) |> liftAsync // Skip at least one frames
+                do! Async.SwitchToContext uiContext |> liftAsync
+                let layer = lineLensLayer
+                // Now relayout the existing adornments
+                if nonVisibleLineNumbers.Count > 0 || newVisibleLineNumbers.Count > 0 then
+                    for lineNumber in visibleLineNumbers do
+                        let line = 
+                            (buffer.GetLineFromLineNumber lineNumber).Start
+                            |> view.GetTextViewLineContainingBufferPosition
+                        applyFuncOnLineStackPanels line (fun ui ->
+                            ui.Visibility <- Visibility.Visible
+                            layoutUIElementOnLine line ui
+                        )
+                do! Async.Sleep(495) |> liftAsync
+                try
+                    for visibleLineNumber in visibleLineNumbers do
+                        if trackingSpans.ContainsKey visibleLineNumber then
+                            trackingSpans.[visibleLineNumber] 
+                            |> Seq.map (fun trackingSpan ->
+                                    let success, res = uiElements.TryGetValue trackingSpan
+                                    if success then res else null
+                                )
+                            |> Seq.filter (isNull >> not)
+                            |> Seq.filter (addedAdornments.Contains >> not)
+                            |> Seq.iter(fun stackPanel ->
+                                    layer.AddAdornment(AdornmentPositioningBehavior.ViewportRelative, Nullable(), 
+                                        self, stackPanel, AdornmentRemovedCallback(fun _ _ -> ())) |> ignore
```

Comment:
Should the callback remove the item from addedAdornments? It's a leak otherwise.

Reply:
It must be removed in the `RemoceCodeLens` method to not spread the necessary steps for removing code lens. 

Many many thanks for telling me this, because this is definitely a memory leak.